ENG-2129: Reduces visual footprint of search bar

vsreekanti opened the pull request:
## Describe your changes and why you are making these changes

This PR reduces how imposing the search bar is in the `PagintedSearchTable` component by making it an underlined text field of a smaller size. It also only uses one input adornment — a search icon if the field is empty and an `X` icon otherwise.

This PR also adds sort functionality to the search table. Please seem the Loom video for more details.

## Related issue number (if any)

ENG-2120

## Loom demo (if any)

https://www.loom.com/share/7b5e2a1a401549548e65dab4c9b38d53

## Checklist before requesting a review
- [x] I have created a descriptive PR title. The PR title should complete the sentence "This PR...".
- [x] I have performed a self-review of my code.
- [x] I have included a small demo of the changes. For the UI, this would be a screenshot or a Loom video.
- [N/A] If this is a new feature, I have added unit tests and integration tests.
- [N/A] I have run the integration tests locally and they are passing.
- [x] I have run the linter script locally (See `python3 scripts/run_linters.py -h` for usage).
- [x] All features on the UI continue to work correctly.
- [x] Added one of the following CI labels:
    - `run_integration_test`: Runs integration tests
    - `skip_integration_test`: Skips integration tests (Should be used when changes are ONLY documentation/UI)

likawind commented on src/ui/common/src/components/tables/PaginatedSearchTable.tsx:
```diff
@@ -225,6 +226,7 @@ export const PaginatedSearchTable: React.FC<PaginatedSearchTableProps> = ({
                                 columnIndex < columns.length - 1
                                   ? '1px solid rgba(224, 224, 224, 1);'
                                   : 'none',
+                              fontSize: '16px', // This is needed for consistency.
```
@agiron123 is there a single place we can track all font sizes? It would make fonts easier to track and less effort keeping them consistent

agiron123 replied:
Yes, theme.tsx should be the place to put this. Ideally we should be using units throughout our project. (and even add things like fontWeight to our theme file).

I am pretty guilty of just using pixel values, when I should be using our theme's unit system instead.

I've filed a task here to address us not using units everywhere in our project:
https://linear.app/aqueducthq/issue/ENG-2236/use-theme-constants-throughout-frontend-project

likawind commented on src/ui/common/src/components/tables/PaginatedSearchTable.tsx:
```diff
@@ -128,38 +128,39 @@ export const PaginatedSearchTable: React.FC<PaginatedSearchTableProps> = ({
             id="outlined-basic"
-            variant="outlined"
-            fullWidth
+            variant="standard"
```
what's the difference between 'outlined' and 'standard'?

vsreekanti commented:
@likawind @agiron123, based on Andre's feedback, I added sort functionality and modified the design. Please see the Loom video above for more details.

agiron123 commented:
> @likawind @agiron123, based on Andre's feedback, I added sort functionality and modified the design. Please see the Loom video above for more details.

@vsreekanti oops, started working on ENG-2105, didn't know that you were going to add sorting here. I've assigned the task to you now instead.
https://linear.app/aqueducthq/issue/ENG-2105/workflows-and-data-list-page-allow-users-to-sort-workflows-by-last

vsreekanti commented:
Ah, sorry for the mix up! You'd tagged me in the mockups on Slack, so I'd assumed you were thinking about this as a part of the search bar work.

agiron123 commented on src/ui/common/src/components/pages/workflows/index.tsx:
```diff
@@ -120,6 +120,25 @@ const WorkflowsPage: React.FC<Props> = ({ user, Layout = DefaultLayout }) => {
+    },
+    {
+      name: 'Status',
+      sortAccessPath: ['name', 'status'],
```
What is this sortAccessPath field? I assume that it allows users to sort by status, but order the results by the name of the item? Or does it alphabetize the statuses?

vsreekanti replied:
It might not be the clearest name, but It's the sequence of keys to use within the data row object to get the field that should be used to sort. It's documented [here](https://github.com/aqueducthq/aqueduct/pull/846/files#diff-f4a84e68fbf99ea9144649623979796ffd0744dd8a42be4c80d8f2cccaa868faR47-R49).

agiron123 commented on src/ui/common/src/components/tables/PaginatedSearchTable.tsx:
```diff
@@ -57,10 +77,18 @@ export const PaginatedSearchTable: React.FC<PaginatedSearchTableProps> = ({
-  let rows = data.data;
-  const columns = data.schema.fields;
+  const [sortMenuAnchor, setSortMenuAnchor] =
+    React.useState<HTMLButtonElement>(null);
```
nit: the rest of our codebase uses just 'useState' or 'useEffect'.

agiron123 commented on src/ui/common/src/components/tables/PaginatedSearchTable.tsx:
```diff
@@ -57,10 +77,18 @@ export const PaginatedSearchTable: React.FC<PaginatedSearchTableProps> = ({
+  const [rows, setRows] = React.useState([...data.data]);
 
-  let filteredRows = [];
+  // let rows = data.data;
```
remove comment.

agiron123 commented on src/ui/common/src/components/tables/PaginatedSearchTable.tsx:
```diff
@@ -125,53 +153,196 @@ export const PaginatedSearchTable: React.FC<PaginatedSearchTableProps> = ({
+
+    const sortedRows = [...rows].sort((r1, r2) => {
+      const col = sortConfig.sortColumn;
+      let v1: PaginatedSearchTableRow | PaginatedSearchTableElement = r1;
```
What do v1 and v2 stand for?

vsreekanti replied:
value1 and value 2. I can update to be clearer.

agiron123 commented on src/ui/common/src/components/tables/PaginatedSearchTable.tsx:
```diff
@@ -125,53 +153,196 @@ export const PaginatedSearchTable: React.FC<PaginatedSearchTableProps> = ({
+        v2 = v2[path];
+      }
+
+      if (sortConfig.sortType === 'asc') {
```
Kind of a silly thing, but can you add a const enum for the sort types? 
That way in the future, we can just say 'SORT_TYPE.ASC' and have stronger typing.

vsreekanti replied:
Yep, makes sense!

agiron123 commented on src/ui/common/src/components/tables/PaginatedSearchTable.tsx:
```diff
@@ -125,53 +153,196 @@ export const PaginatedSearchTable: React.FC<PaginatedSearchTableProps> = ({
+        } else {
+          return 0;
+        }
+      } else {
```
Another way of doing these comparisons is v1.localeCompare(v2) for ascending and v2.localeCompare(v1) for descending.

agiron123 replied:
Either way works though, and may have some caveats for special character handling.

vsreekanti replied:
Neat! That's definitely cleaner — I think it uses the same logic under the hood, so I'll switch it.

vsreekanti replied:
Ah, nevermind. It doesn't work on numbers, only on strings, so this won't work since our element type can have numbers. :l🙁

agiron123 commented on src/ui/common/src/components/tables/PaginatedSearchTable.tsx:
```diff
@@ -125,53 +153,196 @@ export const PaginatedSearchTable: React.FC<PaginatedSearchTableProps> = ({
+              ))}
+            </Menu>
+
+            <Menu
```
This PaginatedSearchTable is starting to get a little long with the search / menu functionality added in. 

Doesn't have to be done right now, but can you add a small refactoring task to turn them into their own respective components?

agiron123 commented on src/ui/common/src/components/tables/PaginatedSearchTable.tsx:
```diff
@@ -125,53 +153,196 @@ export const PaginatedSearchTable: React.FC<PaginatedSearchTableProps> = ({
+              label={`Sort: ${sortConfig.sortColumn.name}`}
+              onDelete={() =>
+                setSortConfig({
+                  sortType: '',
```
Would be nice to have this in a constant, i.e. SORT_TYPE.NONE

agiron123 reviewed (approved):
Nice work here! I really like how this turned out.
